enhancements/network/user-defined-network-for-virt-workloads.md

```diff
@@ -0,0 +1,563 @@
+sequenceDiagram
+  actor user as user
+  participant api-server as api-server
+  participant KubeVirt as KubeVirt
+  participant kubevirt-ipam-extensions as kubevirt-ipam-extensions
+  participant OVN-Kubernetes as OVN-Kubernetes
+
+  user ->>+ api-server: createVM(name=vm-1, ...)
+  api-server -->>- user: OK
+  api-server -)+ KubeVirt: reconcile VM
+  api-server -)+ kubevirt-ipam-extensions: reconcile VM
+  KubeVirt ->>+ api-server: templatePodForVM()
+  api-server -->> KubeVirt: OK
+  KubeVirt -->>- api-server: OK
+  rect rgb(0, 255, 150)
+    Note over kubevirt-ipam-extensions: mutation webhook
+    api-server ->> kubevirt-ipam-extensions: POD
+    kubevirt-ipam-extensions ->> api-server: annotatePod(ipamClaimName)
+    api-server -->>- kubevirt-ipam-extensions: OK
+  end
+  rect rgb(0, 255, 255)
+    Note over kubevirt-ipam-extensions: VM / VMI controller
+    kubevirt-ipam-extensions ->> api-server: createIPAMClaim(ipamClaimName=vm-1.udn)
+    api-server -->> kubevirt-ipam-extensions: OK
+    kubevirt-ipam-extensions -->>- api-server: OK
+  end
+  api-server -) OVN-Kubernetes: reconcile POD
+
+  OVN-Kubernetes -->> api-server: OK
+```
+
+### API Extensions
+
+We will need to extend the `IPAMClaim` CRD to support three new things:
+- allow the user to specify which IP addresses they want to use for their UDNs
+- allow OVN-Kubernetes to report back success / errors on the `IPAMClaim`
+- optional: allow OVN-Kubernetes to report which pod is currently holding the
+claim
+
+The first two items are required, and the latter is an improvement, intended to
+make the solution more transparent / secure. This way, OVN-Kubernetes would
+reject reconciling a claim currently held by another pod (unless this pod
+belongs to the same VM - i.e. migration use case). We will check the owner
+reference of the `IPAMClaim` versus the owner reference of the pod to assure
+this scenario (thus ensuring OVN-Kubernetes doesn't need to understand what a
+VM is to do so).
+
+The proposed updated CRD would look like:
+```go
+// IPAMClaim is the Schema for the IPAMClaim API
+type IPAMClaim struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   IPAMClaimSpec   `json:"spec,omitempty"`
+	Status IPAMClaimStatus `json:"status,omitempty"`
+}
+
+type IPAMClaimSpec struct {
+	// The network name for which this persistent allocation was created
+	Network string `json:"network"`
+	// The pod interface name for which this allocation was created
+	Interface string `json:"interface"` // *new* attribute
+	// The IP requested by the user
+	// +optional
+	IPRequest []net.IP `json:"ipRequest,omitempty"`
+}
+
+// IPAMClaimStatus contains the observed status of the IPAMClaim.
+type IPAMClaimStatus struct {
+	// The list of IP addresses (v4, v6) that were allocated for the pod interface
+	IPs []string `json:"ips"`
+	// The name of the pod holding the IPAMClaim
+	OwnerPod string `json:"ownerPod"`                            // *new* attribute
```

Comment:
We can use a proper `metav1.OwnerRef` object, also don't we need this to be multiple for live migration purposes ?

Reply:
I would rather not involve the scheduler and cascading deletion here. Hence trying to move away from having multiple owner refs.

For live-migration we would check *also* the owner ref, and realize both pods are for the same entity.